13/abidmunirmalik/directors.py

```diff
@@ -0,0 +1,82 @@
+import csv
+from collections import defaultdict, namedtuple, Counter
+
+MOVIE_DATA = 'movie_metadata.csv'
+NUM_TOP_DIRECTORS = 20
+MIN_MOVIES = 4
+MIN_YEAR = 1960
+
+def get_movies_by_director():
+    '''Extracts all movies from csv and stores them in a dictionary
+    where keys are directors, and values is a list of movies (named tuples)'''
+    Movies  = namedtuple('Movies', ['title','year','score'])
+    movies_by_director = defaultdict(list)
+    with open(MOVIE_DATA) as csvfile:
+        reader = csv.DictReader(csvfile)
+        for rows in reader:
+            try:
+                director = rows['director_name']
+                title = rows['movie_title'].replace('\xa0','')
+                year =  int(rows['title_year'])
+                score = float(rows['imdb_score'])
+            except ValueError:
+                continue
+            movie = Movies(title=title, year=year, score=score)
+            movies_by_director[director].append(movie)     
+    return movies_by_director
+
+def filter_directors(directors):
+    directors_with_less_than_min_movies = []
+    for director,movies in directors.items():
+        d,m = director,len(movies)
+        if m < MIN_MOVIES:
+            directors_with_less_than_min_movies.append(d)
+    for _ in directors_with_less_than_min_movies:
+        del directors[_]
+    return directors    
+
+    
+def get_average_scores(directors):
+    '''Filter directors with < MIN_MOVIES and calculate averge score'''
+    rating_ranks = defaultdict(list)
+    for director, movies in directors.items():
+        rating_ranks[director].append(_calc_mean(movies) )
+    mylist = Counter(rating_ranks).most_common(20)
+    return directors,mylist
+
+def _calc_mean(movies):
+    '''Helper method to calculate mean of list of Movie namedtuples'''
+    return round(sum(movie.score for movie in movies) / len(movies),1)
+
+
+def print_results(directors,final_list):
+    '''Print directors ordered by highest average rating. For each director
+    print his/her movies also ordered by highest rated movie.
+    See http://pybit.es/codechallenge13.html for example output'''
+    sep_line = '-' * 60
+    for i in range(len(final_list)):
+        counter = str((i+1)).zfill(2)+ '.'
+        d = final_list[i][0]
```

Comment:
this type of for loop is not common in Python, a more Pythonic way is to use `for d, m in final_list` or if you need the index for printing issues `for i, e in enumerate(final_list): d, av = e` because if e is an tuple with two elements you can use Python's magic unpacking and directly assign the two elements to two variables